File: Java-CodeServer/src/main/resources/js/prompto/grammar/ArgumentAssignment.js
var SyntaxError = require("../error/SyntaxError").SyntaxError;
var ContextualExpression = require("../value/ContextualExpression").ContextualExpression;
var CategoryType = null;
var MemberSelector = require("../expression/MemberSelector").MemberSelector;
var Variable = require("../runtime/Variable").Variable;
var VoidType = require("../type/VoidType").VoidType;

exports.resolve = function() {
    CategoryType = require("../type/CategoryType").CategoryType;
}

function ArgumentAssignment(argument, expression) {
	this.argument = argument;
	this.expression = expression;
	return this;
}

Object.defineProperty(ArgumentAssignment.prototype, "name", {
	get : function() {
		return this.argument.name;
	}
});

ArgumentAssignment.prototype.toDialect = function(writer) {
    writer.toDialect(this);
};

ArgumentAssignment.prototype.toODialect = function(writer) {
    if(this.argument!=null) {
        writer.append(this.argument.name);
        writer.append(" = ");
    }
    this.expression.toDialect(writer);
};

ArgumentAssignment.prototype.toSDialect = function(writer) {
    if(this.argument!=null) {
        writer.append(this.argument.name);
        writer.append(" = ");
    }
    this.expression.toDialect(writer);
};

ArgumentAssignment.prototype.toEDialect = function(writer) {
    this.expression.toDialect(writer);
    if(this.argument!=null) {
        writer.append(" as ");
        writer.append(this.argument.name);
    }
};

ArgumentAssignment.prototype.toString = function() {
    if(this.argument===null) {
        return this.expression.toString();
    } else {
        return this.name + " = " + this.expression.toString();
    }
};

ArgumentAssignment.prototype.equals = function(obj) {
	if(obj==this) {
		return true;
	} else if(obj==null) {
		return false;
	} else if(!(obj instanceof ArgumentAssignment)) {
		return false;
	} else {
		return this.argument.equals(obj.argument) &&
			this.expression.equals(other.expression);
	}
};

ArgumentAssignment.prototype.check = function(context) {
	var actual = context.getRegisteredValue(this.argument.name);
	if(actual==null) {
        var actualType = this.expression.check(context);
		context.registerValue(new Variable(this.argument.name, actualType));
	} else {
		// need to check type compatibility
		var actualType = actual.getType(context);
		var newType = this.expression.check(context);
		newType.checkAssignableTo(context,actualType);
	}
	return VoidType.instance;
};

ArgumentAssignment.prototype.evaluate = function(context) {
	if(context.getRegisteredValue(this.argument.name)==null) {
		context.registerValue(new Variable(this.argument.name, this.expression));
	}
	context.setValue(this.argument.name, this.expression.interpret(context));
	return null;
};

ArgumentAssignment.prototype.resolve = function(context, methodDeclaration, checkInstance) {
	// since we support implicit members, it's time to resolve them
	var name = this.argument.name;
	var expression = this.expression;
	var argument = methodDeclaration.args.find(name);
	var required = argument.getType(context);
	var actual = expression.check(context.getCallingContext());
	if(checkInstance && actual instanceof CategoryType) {
		var value = expression.interpret(context.getCallingContext());
		if(value && value.getType) {
			actual = value.getType();
		}
	}
	if(!actual.isAssignableTo(context, required) && (actual instanceof CategoryType)) {
		expression = new MemberSelector(expression, name);
	}
	return expression;
};

ArgumentAssignment.prototype.makeAssignment = function(context, declaration) {
	var argument = this.argument;
	// when 1st argument, can be unnamed
	if(argument===null) {
		if(declaration.args.length==0) {
			throw new SyntaxError("Method has no argument");
		}
		argument = declaration.args[0];
	} else {
		argument = declaration.args.find(this.name);
	}
	if(argument==null) {
		throw new SyntaxError("Method has no argument:" + this.name);
	}
	var expression = new ContextualExpression(context,this.expression);
	return new ArgumentAssignment(argument,expression);
};

exports.ArgumentAssignment = ArgumentAssignment;

File: Java-Core/src/main/java/prompto/grammar/ArgumentAssignment.java
package prompto.grammar;

import prompto.argument.IArgument;
import prompto.declaration.IMethodDeclaration;
import prompto.error.PromptoError;
import prompto.error.SyntaxError;
import prompto.expression.IExpression;
import prompto.expression.MemberSelector;
import prompto.runtime.Context;
import prompto.runtime.Variable;
import prompto.type.CategoryType;
import prompto.type.IType;
import prompto.type.VoidType;
import prompto.utils.CodeWriter;
import prompto.value.ContextualExpression;
import prompto.value.IInstance;

public class ArgumentAssignment {
	
	IArgument argument;
	IExpression expression;
	
	public ArgumentAssignment(IArgument argument, IExpression expression) {
		this.argument = argument;
		this.expression = expression;
	}
	
	public IArgument getArgument() {
		return argument;
	}

	public Identifier getArgumentId() {
		return argument==null ? null : argument.getId();
	} 
	
	public IExpression getExpression() {
		return expression;
	}

	public void setExpression(IExpression expression) {
		this.expression = expression;
	}
	
	@Override
	public String toString() {
		return (argument!=null ? argument.getId() + " = " : "") + expression.toString();
	}
	
	public void toDialect(CodeWriter writer) {
		switch(writer.getDialect()) {
		case E:
			toEDialect(writer);
			break;
		case O:
			toODialect(writer);
			break;
		case S:
			toSDialect(writer);
			break;
		}
	}
	
	private void toODialect(CodeWriter writer) {
		if(argument!=null) {
			writer.append(argument.getId());
			writer.append(" = ");
		}
		expression.toDialect(writer);
	}

	private void toSDialect(CodeWriter writer) {
		if(argument!=null) {
			writer.append(argument.getId());
			writer.append(" = ");
		}
		expression.toDialect(writer);
	}

	private void toEDialect(CodeWriter writer) {
		expression.toDialect(writer);
		if(argument!=null) {
			writer.append(" as ");
			writer.append(argument.getId());
		}
	}
	
	@Override
	public boolean equals(Object obj) {
		if(obj==this)
			return true;
		if(obj==null)
			return false;
		if(!(obj instanceof ArgumentAssignment))
			return false;
		ArgumentAssignment other = (ArgumentAssignment)obj;
		return this.getArgument().equals(other.getArgument())
				&& this.getExpression().equals(other.getExpression());
	}
	
	public IType check(Context context) {
		INamed actual = context.getRegisteredValue(INamed.class,argument.getId());
		if(actual==null) {
			IType actualType = expression.check(context);
			context.registerValue(new Variable(argument.getId(), actualType));
		} else {
			// need to check type compatibility
			IType actualType = actual.getType(context);
			IType newType = expression.check(context);
			newType.checkAssignableTo(context,actualType);
		}
		return VoidType.instance();
	}
	
	public IExpression resolve(Context context, IMethodDeclaration methodDeclaration,boolean checkInstance) throws PromptoError {
		// since we support implicit members, it's time to resolve them
		Identifier name = argument.getId();
		IExpression expression = getExpression();
		IArgument argument = methodDeclaration.getArguments().find(name);
		IType required = argument.getType(context);
		IType actual = expression.check(context.getCallingContext());
		if(checkInstance && actual instanceof CategoryType) {
			Object value = expression.interpret(context.getCallingContext());
			if(value instanceof IInstance)
				actual = ((IInstance)value).getType();
		}
		if(!actual.isAssignableTo(context, required) && (actual instanceof CategoryType)) 
			expression = new MemberSelector(expression,name);
		return expression; 
	}

	public ArgumentAssignment resolveAndCheck(Context context, ArgumentList argumentList) {
		IArgument argument = this.argument;
		// when 1st argument, can be unnamed
		if(argument==null) {
			if(argumentList.size()==0)
				throw new SyntaxError("Method has no argument");
			argument = argumentList.get(0);
		} else
			argument = argumentList.find(this.getArgumentId());
		if(argument==null)
			throw new SyntaxError("Method has no argument:" + this.getArgumentId());
		IExpression expression = new ContextualExpression(context, this.expression);
		return new ArgumentAssignment(argument,expression);
	}
}
